File: benchmarks/fixtures/python-cross-file-ssrf/src/upstream.py
import requests

MAX_PREVIEW_BYTES = 64 * 1024


def fetch_preview(target):
    with requests.get(target, stream=True, timeout=(2, 2)) as response:
        return read_limited_body(response)


def read_limited_body(response):
    chunks = []
    length = 0
    for chunk in response.iter_content(chunk_size=8 * 1024):
        length += len(chunk)
        if length > MAX_PREVIEW_BYTES:
            raise ValueError("preview response exceeds size limit")
        chunks.append(chunk)
    return b"".join(chunks)

File: benchmarks/fixtures/javascript-cross-file-ssrf/src/upstream.js
const MAX_PREVIEW_BYTES = 64 * 1024;

export async function fetchPreview(target, response) {
  const upstream = await fetch(target, {
    redirect: "follow",
    signal: AbortSignal.timeout(2_000),
  });
  return response.send(await readLimitedBody(upstream));
}

async function readLimitedBody(upstream) {
  if (upstream.body === null) return Buffer.alloc(0);
  const reader = upstream.body.getReader();
  const chunks = [];
  let length = 0;
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;
      length += value.byteLength;
      if (length > MAX_PREVIEW_BYTES) {
        await reader.cancel("preview response exceeds size limit");
        throw new Error("preview response exceeds size limit");
      }
      chunks.push(Buffer.from(value));
    }
  } finally {
    reader.releaseLock();
  }
  return Buffer.concat(chunks, length);
}
